openedx/openedx-platform
# build: move collectstatic ignore patterns into configuration

## Overview

We are [reimplementing edx-platform's static asset processing system](https://github.com/openedx/edx-platform/blob/master/docs/decisions/0017-reimplement-asset-processing.rst). The three main parts are:
1. building
2. collection (prod only)
3. watching (dev only)

This PR addresses (2)

## Description

Django provides the `collectstatic` management command, which collects static assets into the STATIC_ROOT so that they can be served by some system external to Django (like nginx or caddy), as is usually desired in production environments.

edx-platform contains several types of files that we don't want to be collected into the STATIC_ROOT. Previously, these files had to be supplied to the management command using the `--ignore` option:

    ./manage.py lms collectstatic --ignore geoip --ignore sass ...etc
    ./manage.py cms collectstatic --ignore geoip --ignore sass ...etc

This yields a long, hard-to-remember command. Paver wrapped the command in its big `paver update_assets` task, but that task also builds assets, which is totally overkill when you're just trying to collect them.

Fortunately, `collectstatic`'s default ignore patterns can be configured by defining a custom AppConfig class. We define such an config (`EdxPlatformStaticFilesConfig`) for LMS and CMS. Now, devs can collect LMS & CMS assets with just:

    ./manage.py lms collectstatic
    ./manage.py cms collectstatic

Further reading on collecstatic and --ignore:
https://docs.djangoproject.com/en/3.2/ref/contrib/staticfiles/#customizing-the-ignored-pattern-list

Further reading on eschewing Paver:
https://github.com/openedx/edx-platform/blob/master/docs/decisions/0017-reimplement-asset-processing.rst

Closes: https://github.com/openedx/edx-platform/issues/31658

## Testing instructions

Here are steps I used to test. They assume you have Tutor Nightly installed.

* Ensure your Tutor Nightly is up to date and running:
  * In your tutor repo, check out nightly and git-pull.
  * `tutor config save`
  * `tutor images pull all`
  * `tutor local stop && tutor local start -d`
* Generate lists of static files from the LMS & CMS containers and save them to your machine:
  *  `tutor local exec lms find /openedx/staticfiles | sort > staticfiles-lms-master.out`
  *  `tutor local exec cms find /openedx/staticfiles | sort > staticfiles-cms-master.out`
* Confirm both files have 29171 lines:
  *  `wc -l staticfiles-lms-master.out`
  *  `wc -l staticfiles-cms-master.out`
* Rebuild the openedx image using this version of edx-platform:
  *  `tutor config save --set EDX_PLATFORM_REPOSITORY=https://github.com/kdmccormick/edx-platform --set EDX_PLATFORM_VERSION=kdmccormick/assets-collect`
  * `tutor images build openedx`
  * `tutor local stop && tutor local start -d`
* Re-generate lists of static files from the LMS & CMS containers and save them to your machine:
  *  `tutor local exec lms find /openedx/staticfiles | sort > staticfiles-lms-pr.out`
  *  `tutor local exec cms find /openedx/staticfiles | sort > staticfiles-cms-pr.out`
* Confirm that static files lists for LMS and CMS match. These commands should either have no output, or should have a trivial output (for example, the hashes of some assets may have changed; this is OK as long as that for every files missing from one list, it is present with a similar filename in the other list)
  * `diff staticfiles-lms-master.out staticfiles-lms-pr.out`
  * `diff staticfiles-cms-master.out staticfiles-cms-pr.out`
